clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go

```diff
@@ -0,0 +1,283 @@
+// Services can be specified by field and/or label selectors.
+// If namespace is not passed by parameter, all LoadBalancer type service with all-namespace scope is assumed.
+// "measure" action triggers nodesync and observation of nodesync completion for selected LB services.
+// "gather" returns node sync latency summary.
+func (s *LoadBalancerNodeSyncMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	s.client = config.ClusterFramework.GetClientSets().GetClient()
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	switch action {
+	case "measure":
+		if err := s.selector.Parse(config.Params); err != nil {
+			return nil, err
+		}
+		s.waitTimeout, err = util.GetDurationOrDefault(config.Params, "waitTimeout", defaultNodeSyncLatencyTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, s.measureNodeSyncLatency()
+	case "gather":
+		if err := s.labelNodeForLBs(false); err != nil {
+			return nil, err
+		}
+		return s.gather(config.Identifier)
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (s *LoadBalancerNodeSyncMeasurement) Dispose() {}
+
+func (s *LoadBalancerNodeSyncMeasurement) String() string {
+	return loadBalancerNodeSyncLatencyName + ": " + s.selector.String()
+}
+
+func (s *LoadBalancerNodeSyncMeasurement) measureNodeSyncLatency() error {
+	ctx := context.Background()
+	options := metav1.ListOptions{}
+	s.selector.ApplySelectors(&options)
+	svcList, err := s.client.CoreV1().Services(s.selector.Namespace).List(ctx, options)
+	if err != nil {
+		return err
+	}
+
+	// lbSvcMap contains key->service mapping for load balancer type services
+	lbSvcMap := map[string]v1.Service{}
+	for _, svc := range svcList.Items {
+		if svc.Spec.Type == v1.ServiceTypeLoadBalancer {
+			key, err := runtimeobjects.CreateMetaNamespaceKey(&svc)
+			if err != nil {
+				return fmt.Errorf("meta key created error: %v", err)
+			}
+
+			lbSvcMap[key] = svc
+		}
+	}
+	totalLbSvc := len(lbSvcMap)
+
+	// trigger node sync by picking a node and add exclude lb label
+	nodeList, err := s.client.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, node := range nodeList.Items {
+		if !isCandidateNode(node) {
+			continue
+		}
+		s.excludedNodeName = node.Name
+	}
+
+	if s.excludedNodeName == "" {
+		return fmt.Errorf("failed to find a node candidate to trigger nodesync from node list: %v", nodeList.Items)
+	}
+	if err = s.labelNodeForLBs(true); err != nil {
+		return err
+	}
+
+	now := time.Now()
+	for key := range lbSvcMap {
+		s.svcNodeSyncLatencyTracker.Set(key, phaseNodeSyncStart, now)
+	}
+
+	return wait.Poll(5*time.Second, s.waitTimeout, func() (bool, error) {
+		for key, svc := range lbSvcMap {
+			if _, found := s.svcNodeSyncLatencyTracker.Get(key, phaseNodeSyncComplete); found {
+				continue
+			}
+
+			eventList, err := s.client.CoreV1().Events(s.selector.Namespace).Search(Scheme, &svc)
```

Comment:
I'm not a fan of depending on events in a programatic way - events aren't guaranteed.

What if instead we would make the measurement GCE-specific and try to get TargetPool contents from GCE and detect if the node was added/removed based on this?

The event-based test is really implementation-specific - we should target something more blackbox.

Reply:
Making a measurement using GCE API calls has a few downsides in this tests. The test has to loop thru a potential high number of LBs and make GCE API calls repeatedly until the expected state has reached. This would eat up the read RPS quota in GCE the project and potentially cause controllers to hit GCE RateLimiting errors. Having a higher interval would make the measurement less accurate when high number of LBs are used in the load test.


Yes. I agree depending on events are less straightforward. We brainstormed a few ways to make this measurement. One better alternative is to read a metrics exposed by kube-controller-manager directly on master, but cluster loader currently cannot reach master. While the event creation timestamp does provide an accurate latency measurement and it is more a blackbox test than checking TargetPool.